### PR TITLE
Better Update Mask support

### PR DESCRIPTION
**Pull request by rambleraptor**
-----------------------------------------------------------------
# [all]
Native Update Mask support

### CODE REVIEW

File: api/resource.rb [hunk 1]
```diff
@@ -357,6 +357,15 @@ def regex_url
                    .gsub('{{zone}}', '[a-z1-9\-]*')
     end
 
+    # All setable properties in the resource.
```
Comment (ndmckinley) [inline api/resource.rb]: typo

File: api/resource.rb [hunk 2]
```diff
@@ -357,6 +357,15 @@ def regex_url
...
+    # Fingerprints aren't *really" settable properties, but they behave like one.
+    # At Create, they have no value but they can just be read in anyways, and after a Read
+    # they will need ot be set in every Update.
+    def settable_properties
+      all_user_properties.reject { |v| v.output && !v.is_a?(Api::Type::Fingerprint) }
```
Comment (ndmckinley) [inline api/resource.rb]: So to be clear we *do* want to include fingerprints?

File: templates/terraform/resource.erb [hunk 3]
```diff
@@ -22,11 +22,9 @@ package google
     resource_name = product_ns + object.name
     properties = object.all_user_properties
     # Fingerprints aren't *really* settable properties, but they behave like one. At Create, they have no value but they
```
Comment (ndmckinley) [inline templates/terraform/resource.erb]: Can remove comment now.

File: templates/terraform/resource.erb [hunk 4]
```diff
@@ -382,6 +380,7 @@ func resource<%= resource_name -%>Update(d *schema.ResourceData, meta interface{
     }
 
     log.Printf("[DEBUG] Updating <%= object.name -%> %q: %#v", d.Id(), obj)
+<%= lines(compile('templates/terraform/pre_update/update_mask.erb')) if object.all_user_properties.select { |v| v.is_a?(Api::Type::UpdateMask) }.first -%>
```
Comment (ndmckinley) [inline templates/terraform/resource.erb]: Ideal to move it - `pre_update` is where we store custom code, which this no longer is.

File: api/resource.rb (hunk 2 above)
Comment (rileykarson) [reply api/resource.rb]: Yes! That change was me, I forget the _exact_ reasoning but fingerprint acted like a settable/updatable field at every time but `Create`.

**Comment by rambleraptor** [comment]
Update Masks on Ansible now! I had to get Redis working on Ansible or else all of this code would be dead.

PTAL.

File: products/redis/api.yaml [hunk 5]
```diff
@@ -172,3 +172,7 @@ objects:
           - :STANDARD_HA
         default_value: :BASIC
         input: true
+      - !ruby/object:Api::Type::UpdateMask
+        name: updateMask
+        description: The update mask (internal only)
+        output: true
```
Comment (ndmckinley) [inline products/redis/api.yaml]: What is an output only update mask?

Comment (rambleraptor) [reply products/redis/api.yaml]: I didn't know of a great way to describe this.

Do we place updateMask as a property or as a flag on a resource?

If it's a property, it's a property that is neither user-settable nor outputted  in any meaningful way.

Comment (ndmckinley) [reply products/redis/api.yaml]: Right.  Personally, I think it's a flag on a resource, but I wasn't going to quibble about the representation - there's precedent for both.

Comment (rambleraptor) [reply products/redis/api.yaml]: I'll make it a flag on a resource then.

I was worried about the case where an "updateMask" wasn't actually called "updateMask". In that case, we could add more flags to allow us to override the name...or we could just treat it as a property.

Comment (rileykarson) [reply products/redis/api.yaml]: drive-by thoughts:

* We probably don't need to futureproof- a name change will probably come with a behaviour change too, invalidating what we have anyways.

* If we do want to regardless, we could always make it a string-type field where the user provides what the field is called.

**Comment by ndmckinley** [comment]
Oh, I didn't remember that there were terraform downstreams.  That's the normal behavior of the magician, though.

**Comment by ndmckinley** [comment]
Oh, they look good to me, though. @rambleraptor can you add them back in and merge them also?

**Comment by rambleraptor** [comment]
Can I get someone to merge in the TF downstreams?

**Comment by ndmckinley** [comment]
merged em!

**Pull request merged**